src/Stratis.Bitcoin/P2P/PeerSelector.cs

```diff
@@ -82,25 +88,27 @@ private IEnumerable<PeerAddress> SelectPreferredPeers()
 
             // First check to see if there are handshaked peers. If so,
             // give them a 50% chance to be picked over all the other peers.
-            if (this.peerAddresses.Handshaked().Any())
+            var handshaked = this.peerAddresses.Handshaked();
+            if (handshaked.Any())
             {
                 int chance = this.random.Next(100);
                 if (chance <= 50)
```

Comment:
I admit, I don't quite understand the 50% chance logic.  Isn't the preferred order simply handshaked, connected, fresh, attempted, as in, if you have handshaked peers then are they not preferable to anything else?  I remember us talking about handshaked vs lastseen - have they been conflated now to a single timetamp?

Reply:
Hi @345paul! Ok so this method is used only when we want to connect to a single peer during normal node operation. This is not used during PeerDiscovery or GetAddrPayload. 

The idea here IS to prefer peers in the order you specified, namely, handshaked, connected, fresh and attempted. I.e first check to see if there are handshaked peers, if so, give them a 50% chance to be picked. If this fails, check if there are "connected only" peers, if so, give them a 50% chance to be picked. If all else fails pick an attempted or fresh peer randomly

The idea here is that we cant JUST connect to "good" peers. We also have to give fresh or previously attempted peers another chance. I.e. fresh peers that are potentially "good" must also be given a chance to be propagated through the network. Hence, the 50% chance given to handshaked and/or connected peers.

Reply:
@345paul the "LastSeen" rename of "LastConnectedHandshake" and its subsequently "live" updating will come after this PR is merged ;)